File: spikes/m0_steam_spike.py
"""M0 spike — preuve de la chaîne Steam (jetable, aucune app).

Usage:
    uv run spikes/m0_steam_spike.py

Lit STEAM_API_KEY et STEAM_ID depuis l'environnement ou le fichier .env
à la racine du repo. Ne construit rien : mesure, diagnostique, tranche.
"""

# /// script
# requires-python = ">=3.12"
# dependencies = ["httpx>=0.27"]
# ///

from __future__ import annotations

import os
import sys
import time
from pathlib import Path

import httpx

ROOT = Path(__file__).resolve().parent.parent
API = "https://api.steampowered.com"
STORE = "https://store.steampowered.com"
TIMEOUT = 20.0
PACE_S = 1.5
FALLBACK_APPDETAILS_LIMIT = 5


def load_env() -> None:
    env_file = ROOT / ".env"
    if not env_file.exists():
        return
    for raw in env_file.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, _, value = line.partition("=")
        os.environ.setdefault(key.strip(), value.strip())


def get_json(client: httpx.Client, url: str, params: dict[str, object]) -> httpx.Response:
    """GET avec gestion 429 / backoff. follow_redirects=True (piège httpx)."""
    delay = 2.0
    for attempt in range(4):
        resp = client.get(url, params=params, follow_redirects=True)
        if resp.status_code == 429:
            print(f"  [429] rate limit, backoff {delay:.0f}s (tentative {attempt + 1}/4)")
            time.sleep(delay)
            delay *= 2
            continue
        return resp
    return resp


def main() -> int:
    load_env()
    key = os.environ.get("STEAM_API_KEY", "").strip()
    steam_id = os.environ.get("STEAM_ID", "").strip()
    if not steam_id:
        print("[ERREUR] STEAM_ID absent (.env ou environnement).")
        return 2
    print(f"SteamID : {steam_id}")
    if not key:
        print("[ERREUR] STEAM_API_KEY absente : GetOwnedGames est impossible.")
        print("  -> https://steamcommunity.com/dev/apikey puis remplir .env")
        return 2

    with httpx.Client(timeout=TIMEOUT, headers={"User-Agent": "gameshelf-m0-spike"}) as client:
        print("\n== 1. IPlayerService/GetOwnedGames ==")
        t0 = time.perf_counter()
        resp = get_json(
            client,
            f"{API}/IPlayerService/GetOwnedGames/v1/",
            {
                "key": key,
                "steamid": steam_id,
                "include_appinfo": 1,
                "include_played_free_games": 1,
            },
        )
        elapsed = time.perf_counter() - t0
        print(f"HTTP {resp.status_code} en {elapsed:.2f}s, {len(resp.content)} octets")
        ctype = resp.headers.get("content-type", "")
        if "json" not in ctype:
            print(f"[ERREUR] content-type={ctype!r} — réponse non-JSON (redirection non suivie ?)")
            return 1
        try:
            payload = resp.json()
        except ValueError:
            print("[ERREUR] corps illisible en JSON.")
            return 1

        if resp.status_code == 403:
            print("[ERREUR] 403 : clé Web API invalide ou révoquée.")
            return 1
        if resp.status_code != 200:
            print(f"[ERREUR] HTTP {resp.status_code}: {resp.text[:300]}")
            return 1

        inner = payload.get("response", {})
        eresult = inner.get("eresult", 1)
        if eresult in (25, 84):
            print(f"[ERREUR] x-eresult={eresult} : rate limit / accès refusé. Ralentis et réessaie.")
            return 1
        games = inner.get("games")

        if not games:
            print("\n[DIAGNOSTIC] `games` est VIDE (pas d'erreur HTTP).")
            print("Cause la plus probable : les « Détails du jeu » du profil ne sont pas publics.")
            print("L'API renvoie alors une réponse valide mais vide, sans code d'erreur.")
            print(
                "Correctif : Profil > Modifier > Paramètres de confidentialité >\n"
                "« Détails du jeu » = Public, puis relancer ce script."
            )
            print("\nVERDICT : NO-GO (données indisponibles tant que la privacy n'est pas réglée)")
            return 3

        count = int(inner.get("game_count", len(games)))
        total_min = sum(int(g.get("playtime_forever", 0)) for g in games)
        print(f"[OK] {count} jeux retournés, {total_min} min cumulées ({total_min / 60:.1f} h)")

        top = sorted(games, key=lambda g: int(g.get("playtime_forever", 0)), reverse=True)
        print(f"\n== 2. Enrichissement appdetails des {FALLBACK_APPDETAILS_LIMIT} jeux les plus joués ==")
        hits = 0
        for game in top[:FALLBACK_APPDETAILS_LIMIT]:
            appid = int(game["appid"])
            name = game.get("name", "?")
            playtime_h = int(game.get("playtime_forever", 0)) / 60
            resp2 = get_json(client, f"{STORE}/api/appdetails", {"appids": appid})
            header_image = None
            if resp2.status_code == 200 and "json" in resp2.headers.get("content-type", ""):
                node = resp2.json().get(str(appid), {})
                if node.get("success"):
                    header_image = node.get("data", {}).get("header_image")
            ok = "OK" if header_image else "MANQUANT"
            if header_image:
                hits += 1
            print(f"  [{ok:8}] {appid:>7}  {name[:40]:<40} {playtime_h:7.1f} h  {header_image or ''}")
            time.sleep(PACE_S)

        print("\n== 3. GetRecentlyPlayedGames ==")
        resp3 = get_json(
            client,
            f"{API}/IPlayerService/GetRecentlyPlayedGames/v1/",
            {"key": key, "steamid": steam_id, "count": 10},
        )
        if resp3.status_code == 200 and "json" in resp3.headers.get("content-type", ""):
            recent = resp3.json().get("response", {})
            rcount = int(recent.get("total_count", 0))
            print(f"[OK] {rcount} jeux joués ces 2 dernières semaines")
            for g in recent.get("games", []):
                print(f"  {g.get('name', '?'):<40} 2w={int(g.get('playtime_2weeks', 0)) / 60:.1f} h")
        else:
            print(f"[WARN] GetRecentlyPlayedGames indisponible (HTTP {resp3.status_code})")

        print("\n" + "=" * 60)
        print(f"VERDICT : GO")
        print(f"  - GetOwnedGames  : {count} jeux (compte cohérent avec ~80 attendus)")
        print(f"  - appdetails     : {hits}/{FALLBACK_APPDETAILS_LIMIT} header_image récupérées")
        print("  - privacy profil : Détails du jeu publics confirmés")
        print("=" * 60)
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: spikes/m0_steam_spike.ts
#!/usr/bin/env bun
/**
 * M0' — preuve de la chaîne Steam sur Bun (script jetable, aucune app).
 *
 * Usage:
 *   bun run spikes/m0_steam_spike.ts
 *
 * Bun charge automatiquement .env à la racine du projet.
 * Ne construit rien : mesure, diagnostique, tranche.
 */

const API = "https://api.steampowered.com";
const STORE = "https://store.steampowered.com";
const PACE_MS = 1500;
const LIMIT = 5;

type OwnedGame = {
  appid: number;
  name?: string;
  playtime_forever?: number;
};

type OwnedGamesResponse = {
  response?: {
    game_count?: number;
    games?: OwnedGame[];
    eresult?: number;
  };
};

type AppDetailsNode = {
  success?: boolean;
  data?: { header_image?: string };
};

type RecentGamesResponse = {
  response?: {
    total_count?: number;
    games?: { name?: string; playtime_2weeks?: number }[];
  };
};

const key = (process.env.STEAM_API_KEY ?? "").trim();
const steamId = (process.env.STEAM_ID ?? "").trim();

if (!steamId) {
  console.error("[ERREUR] STEAM_ID absent (.env ou environnement).");
  process.exit(2);
}
console.log(`SteamID : ${steamId}`);
if (!key) {
  console.error("[ERREUR] STEAM_API_KEY absente : GetOwnedGames est impossible.");
  console.error("  -> https://steamcommunity.com/dev/apikey puis remplir .env");
  process.exit(2);
}

async function getJson(base: string, params: Record<string, string | number>): Promise<Response> {
  const url = new URL(base);
  for (const [k, v] of Object.entries(params)) {
    url.searchParams.set(k, String(v));
  }
  let delay = 2000;
  let res: Response | undefined;
  for (let attempt = 0; attempt < 4; attempt++) {
    res = await fetch(url, {
      redirect: "follow",
      headers: { "User-Agent": "gameshelf-m0-spike" },
    });
    if (res.status === 429) {
      console.log(`  [429] rate limit, backoff ${delay / 1000}s (tentative ${attempt + 1}/4)`);
      await Bun.sleep(delay);
      delay *= 2;
      continue;
    }
    return res;
  }
  return res as Response;
}

async function main(): Promise<number> {
  console.log("\n== 1. IPlayerService/GetOwnedGames ==");
  const t0 = performance.now();
  const res = await getJson(`${API}/IPlayerService/GetOwnedGames/v1/`, {
    key,
    steamid: steamId,
    include_appinfo: 1,
    include_played_free_games: 1,
  });
  const elapsed = ((performance.now() - t0) / 1000).toFixed(2);
  const body = await res.text();
  console.log(`HTTP ${res.status} en ${elapsed}s, ${body.length} octets`);

  const ctype = res.headers.get("content-type") ?? "";
  if (!ctype.includes("json")) {
    console.error(`[ERREUR] content-type=${ctype} — réponse non-JSON (redirection non suivie ?)`);
    return 1;
  }
  if (res.status === 403) {
    console.error("[ERREUR] 403 : clé Web API invalide ou révoquée.");
    return 1;
  }
  if (res.status !== 200) {
    console.error(`[ERREUR] HTTP ${res.status}: ${body.slice(0, 300)}`);
    return 1;
  }

  let payload: OwnedGamesResponse;
  try {
    payload = JSON.parse(body) as OwnedGamesResponse;
  } catch {
    console.error("[ERREUR] corps illisible en JSON.");
    return 1;
  }

  const inner = payload.response ?? {};
  const eresult = inner.eresult ?? 1;
  if (eresult === 25 || eresult === 84) {
    console.error(
      `[ERREUR] x-eresult=${eresult} : rate limit / accès refusé. Ralentis et réessaie.`,
    );
    return 1;
  }

  const games = inner.games ?? [];
  if (games.length === 0) {
    console.log("\n[DIAGNOSTIC] `games` est VIDE (pas d'erreur HTTP).");
    console.log("Cause la plus probable : les « Détails du jeu » du profil ne sont pas publics.");
    console.log("L'API renvoie alors une réponse valide mais vide, sans code d'erreur.");
    console.log(
      "Correctif : Profil > Modifier > Paramètres de confidentialité >\n" +
        "« Détails du jeu » = Public, puis relancer ce script.",
    );
    console.log("\nVERDICT : NO-GO (données indisponibles tant que la privacy n'est pas réglée)");
    return 3;
  }

  const count = inner.game_count ?? games.length;
  const totalMin = games.reduce((sum, g) => sum + (g.playtime_forever ?? 0), 0);
  console.log(
    `[OK] ${count} jeux retournés, ${totalMin} min cumulées (${(totalMin / 60).toFixed(1)} h)`,
  );

  const top = [...games].sort((a, b) => (b.playtime_forever ?? 0) - (a.playtime_forever ?? 0));
  console.log(`\n== 2. Enrichissement appdetails des ${LIMIT} jeux les plus joués ==`);
  let hits = 0;
  for (const game of top.slice(0, LIMIT)) {
    const appid = game.appid;
    const name = game.name ?? "?";
    const hours = (game.playtime_forever ?? 0) / 60;
    const res2 = await getJson(`${STORE}/api/appdetails`, { appids: appid });
    let headerImage: string | null = null;
    const ctype2 = res2.headers.get("content-type") ?? "";
    if (res2.status === 200 && ctype2.includes("json")) {
      const node = (JSON.parse(await res2.text()) as Record<string, AppDetailsNode>)[String(appid)];
      if (node?.success) {
        headerImage = node.data?.header_image ?? null;
      }
    }
    if (headerImage) hits++;
    const ok = headerImage ? "OK" : "MANQUANT";
    console.log(
      `  [${ok.padEnd(8)}] ${String(appid).padStart(7)}  ${name.slice(0, 40).padEnd(40)} ${hours.toFixed(1).padStart(7)} h  ${headerImage ?? ""}`,
    );
    await Bun.sleep(PACE_MS);
  }

  console.log("\n== 3. GetRecentlyPlayedGames ==");
  const res3 = await getJson(`${API}/IPlayerService/GetRecentlyPlayedGames/v1/`, {
    key,
    steamid: steamId,
    count: 10,
  });
  const ctype3 = res3.headers.get("content-type") ?? "";
  if (res3.status === 200 && ctype3.includes("json")) {
    const recent = (JSON.parse(await res3.text()) as RecentGamesResponse).response ?? {};
    console.log(`[OK] ${recent.total_count ?? 0} jeux joués ces 2 dernières semaines`);
    for (const g of recent.games ?? []) {
      console.log(
        `  ${(g.name ?? "?").padEnd(40)} 2w=${((g.playtime_2weeks ?? 0) / 60).toFixed(1)} h`,
      );
    }
  } else {
    console.log(`[WARN] GetRecentlyPlayedGames indisponible (HTTP ${res3.status})`);
  }

  console.log(`\n${"=".repeat(60)}`);
  console.log("VERDICT : GO");
  console.log(`  - GetOwnedGames  : ${count} jeux (compte cohérent avec ~80 attendus)`);
  console.log(`  - appdetails     : ${hits}/${LIMIT} header_image récupérées`);
  console.log("  - privacy profil : Détails du jeu publics confirmés");
  console.log("=".repeat(60));
  return 0;
}

process.exit(await main());
